File: strands-py/src/strands/storage/s3_storage.py
"""Amazon S3 storage implementation."""

from __future__ import annotations

import asyncio
import builtins
from typing import Any

from ..types.exceptions import StorageError
from .storage import _NamespacedStorage, _normalize_key, _normalize_prefix

_S3_PAGE_SIZE = 1000


class S3Storage:
    """Persists bytes as objects in an Amazon S3 bucket.

    The AWS SDK (boto3) is imported lazily on first use so applications that
    never use S3 don't pay the import cost.

    Example:
        ```python
        from strands.storage import S3Storage

        storage = S3Storage("my-bucket", prefix="agents/")
        await storage.write("session/abc/state.json", data)
        ```
    """

    def __init__(
        self,
        bucket: str,
        *,
        prefix: str = "",
        region_name: str | None = None,
        boto_session: Any = None,
        boto_client_config: Any = None,
    ) -> None:
        """Initialize S3 storage.

        Args:
            bucket: S3 bucket name.
            prefix: Key prefix prepended to every key (namespace within the bucket).
            region_name: AWS region override.
            boto_session: Pre-configured boto3 session. Cannot combine with region_name.
            boto_client_config: Botocore Config object for the S3 client.

        Raises:
            StorageError: If both region_name and boto_session are provided.
        """
        if region_name is not None and boto_session is not None:
            raise StorageError("Cannot specify both region_name and boto_session")

        self._bucket = bucket
        normalized = _normalize_prefix(prefix)
        self._prefix = f"{normalized}/" if normalized else ""
        self._region_name = region_name
        self._boto_session = boto_session
        self._boto_client_config = boto_client_config
        self._client: Any = None

    async def write(self, key: str, data: bytes) -> None:
        """Store data as an S3 object.

        Args:
            key: Opaque string key identifying the value.
            data: Raw bytes to persist.

        Raises:
            StorageError: If the write fails.
        """
        normalized = _normalize_key(key)
        client = self._get_client()
        object_key = f"{self._prefix}{normalized}"

        try:
            await asyncio.to_thread(client.put_object, Bucket=self._bucket, Key=object_key, Body=data)
        except Exception as error:
            raise StorageError(f"Failed to write '{key}' to S3") from error

    async def read(self, key: str) -> bytes | None:
        """Read an S3 object.

        Args:
            key: The key to read.

        Returns:
            The object contents as bytes, or None if the key does not exist.

        Raises:
            StorageError: If the read fails for a reason other than a missing key.
        """
        normalized = _normalize_key(key)
        client = self._get_client()
        object_key = f"{self._prefix}{normalized}"

        try:
            response = await asyncio.to_thread(client.get_object, Bucket=self._bucket, Key=object_key)
            return await asyncio.to_thread(response["Body"].read)
        except client.exceptions.NoSuchKey:
            return None
        except Exception as error:
            resp = getattr(error, "response", None)
            if resp and resp.get("Error", {}).get("Code") == "NoSuchKey":
                return None
            raise StorageError(f"Failed to read '{key}' from S3") from error

    async def delete(self, key: str) -> None:
        """Delete an S3 object. No-op if the key does not exist.

        Args:
            key: The key to delete.

        Raises:
            StorageError: If the delete fails.
        """
        normalized = _normalize_key(key)
        client = self._get_client()
        object_key = f"{self._prefix}{normalized}"

        try:
            await asyncio.to_thread(client.delete_object, Bucket=self._bucket, Key=object_key)
        except Exception as error:
            raise StorageError(f"Failed to delete '{key}' from S3") from error

    async def list(self, query: str = "") -> builtins.list[str]:
        """List S3 objects matching the given prefix.

        Paginates automatically for large result sets.

        Args:
            query: A prefix string to filter keys. Empty string matches all.

        Returns:
            Matching keys sorted ascending, with the storage-level prefix stripped.

        Raises:
            StorageError: If the listing fails.
        """
        prefix = _normalize_prefix(query)
        client = self._get_client()
        s3_prefix = f"{self._prefix}{prefix}"

        try:
            return await asyncio.to_thread(self._list_sync, client, s3_prefix)
        except Exception as error:
            raise StorageError(f"Failed to list keys with prefix '{query}' from S3") from error

    def _list_sync(self, client: Any, s3_prefix: str) -> builtins.list[str]:
        """Paginate list_objects_v2 synchronously (called via to_thread)."""
        keys: builtins.list[str] = []
        continuation_token: str | None = None

        while True:
            kwargs: dict[str, Any] = {
                "Bucket": self._bucket,
                "Prefix": s3_prefix,
                "MaxKeys": _S3_PAGE_SIZE,
            }
            if continuation_token:
                kwargs["ContinuationToken"] = continuation_token

            response = client.list_objects_v2(**kwargs)

            for obj in response.get("Contents", []):
                key = obj["Key"]
                if self._prefix and key.startswith(self._prefix):
                    key = key[len(self._prefix) :]
                keys.append(key)

            if not response.get("IsTruncated"):
                break
            continuation_token = response.get("NextContinuationToken")

        return sorted(keys)

    def namespace(self, prefix: str) -> _NamespacedStorage:
        """Return a view of this storage with all keys prefixed.

        Args:
            prefix: Prefix to prepend to all keys.

        Returns:
            A namespaced storage view.
        """
        return _NamespacedStorage(self, prefix)

    def _get_client(self) -> Any:
        """Lazily create and cache the S3 client."""
        if self._client is not None:
            return self._client

        import boto3
        from botocore.config import Config

        config = self._boto_client_config
        if config is None:
            config = Config(user_agent_extra="strands-agents")
        elif not getattr(config, "user_agent_extra", None):
            config = config.merge(Config(user_agent_extra="strands-agents"))

        if self._boto_session is not None:
            session = self._boto_session
        else:
            session = boto3.Session(region_name=self._region_name)

        self._client = session.client("s3", config=config)
        return self._client

File: strands-ts/src/storage/s3-storage.ts
import type { Storage } from './storage.js'

import { StorageError } from '../errors.js'
import { namespace, normalizeKey, normalizePrefix } from './storage.js'

/** Configuration for {@link S3Storage}. */
export interface S3StorageConfig {
  /** Optional key prefix prepended to every key (a leading namespace within the bucket). */
  prefix?: string
  /** AWS region override. When omitted, the SDK's standard resolution chain applies. Cannot be combined with `s3Client`. */
  region?: string
  /** Pre-configured S3 client. Cannot be combined with `region`. */
  s3Client?: import('@aws-sdk/client-s3').S3Client
}

const S3_PAGE_SIZE = 1000

/**
 * Amazon S3 {@link Storage} backend.
 *
 * Stores each key as an S3 object under an optional prefix. The AWS SDK is loaded
 * lazily on first use and declared as an optional peer dependency, so consumers that
 * never construct an `S3Storage` are not required to install `@aws-sdk/client-s3`.
 *
 * @example
 * ```typescript
 * import { S3Storage } from '@strands-agents/sdk/storage'
 *
 * const storage = new S3Storage('my-bucket', { prefix: 'agents/' })
 * await storage.write('sessions/abc/snapshot.json', bytes)
 * ```
 */
export class S3Storage implements Storage {
  private readonly _bucket: string
  private readonly _prefix: string
  private readonly _region: string | undefined
  private _client: import('@aws-sdk/client-s3').S3Client | undefined

  /**
   * @param bucket - Target S3 bucket name
   * @param config - Optional prefix, region, or pre-configured client
   * @throws {@link StorageError} if both `region` and `s3Client` are provided
   */
  constructor(bucket: string, config?: S3StorageConfig) {
    if (config?.s3Client && config.region) {
      throw new StorageError('Cannot specify both s3Client and region. Configure the region on the S3Client instead.')
    }
    this._bucket = bucket
    this._prefix = config?.prefix ? config.prefix.split('/').filter(Boolean).join('/') + '/' : ''
    this._region = config?.region
    this._client = config?.s3Client
  }

  /**
   * Stores `data` under `key`, overwriting any existing value.
   *
   * @param key - Opaque string key identifying the value
   * @param data - Raw bytes to persist
   * @throws {@link StorageError} if the key is invalid or the upload fails
   */
  async write(key: string, data: Uint8Array): Promise<void> {
    const normalized = normalizeKey(key)
    const client = await this._getClient()
    const { PutObjectCommand } = await import('@aws-sdk/client-s3')
    try {
      await client.send(new PutObjectCommand({ Bucket: this._bucket, Key: this._objectKey(normalized), Body: data }))
    } catch (error: unknown) {
      throw new StorageError(`Failed to write '${normalized}' to S3 bucket '${this._bucket}'`, { cause: error })
    }
  }

  /**
   * Retrieves the bytes previously stored under `key`.
   *
   * @param key - The key to read
   * @returns The stored bytes, or `null` if no value exists for `key`
   * @throws {@link StorageError} if the key is invalid or the download fails
   */
  async read(key: string): Promise<Uint8Array | null> {
    const normalized = normalizeKey(key)
    const client = await this._getClient()
    const { GetObjectCommand } = await import('@aws-sdk/client-s3')
    try {
      const response = await client.send(
        new GetObjectCommand({ Bucket: this._bucket, Key: this._objectKey(normalized) })
      )
      const body = await response.Body?.transformToByteArray()
      return body ? new Uint8Array(body) : null
    } catch (error: unknown) {
      if (error instanceof Error && (error.name === 'NoSuchKey' || error.name === 'NotFound')) {
        return null
      }
      throw new StorageError(`Failed to read '${normalized}' from S3 bucket '${this._bucket}'`, { cause: error })
    }
  }

  /**
   * Deletes the value stored under `key`. A no-op if the key does not exist.
   *
   * @param key - The key to delete
   * @throws {@link StorageError} if the key is invalid or the delete request fails
   */
  async delete(key: string): Promise<void> {
    const normalized = normalizeKey(key)
    const client = await this._getClient()
    const { DeleteObjectCommand } = await import('@aws-sdk/client-s3')
    try {
      await client.send(new DeleteObjectCommand({ Bucket: this._bucket, Key: this._objectKey(normalized) }))
    } catch (error: unknown) {
      throw new StorageError(`Failed to delete '${normalized}' from S3 bucket '${this._bucket}'`, { cause: error })
    }
  }

  /**
   * Lists the keys whose names begin with `prefix`, sorted lexicographically.
   *
   * @param prefix - Key prefix to match. An empty string matches all keys.
   * @returns The matching keys, sorted ascending
   * @throws {@link StorageError} if the prefix is invalid or the list request fails
   */
  async list(prefix: string): Promise<string[]> {
    const normalized = normalizePrefix(prefix)
    const client = await this._getClient()
    const { ListObjectsV2Command } = await import('@aws-sdk/client-s3')
    const listPrefix = `${this._prefix}${normalized}`
    const keys: string[] = []
    let continuationToken: string | undefined
    try {
      do {
        const response = await client.send(
          new ListObjectsV2Command({
            Bucket: this._bucket,
            Prefix: listPrefix,
            MaxKeys: S3_PAGE_SIZE,
            ContinuationToken: continuationToken,
          })
        )
        for (const object of response.Contents ?? []) {
          if (object.Key === undefined) continue
          keys.push(this._prefix ? object.Key.slice(this._prefix.length) : object.Key)
        }
        continuationToken = response.IsTruncated ? response.NextContinuationToken : undefined
      } while (continuationToken)
    } catch (error: unknown) {
      throw new StorageError(`Failed to list S3 bucket '${this._bucket}' under '${normalized}'`, { cause: error })
    }
    return keys.sort()
  }

  private async _getClient(): Promise<import('@aws-sdk/client-s3').S3Client> {
    if (this._client) return this._client
    const { S3Client } = await import('@aws-sdk/client-s3')
    this._client = new S3Client(this._region ? { region: this._region } : {})
    return this._client
  }

  /** Returns a prefixed view of this storage without mutating the original. */
  namespace(prefix: string): Storage {
    return namespace(this, prefix)
  }

  private _objectKey(key: string): string {
    return `${this._prefix}${key}`
  }
}
